File: 15.js
// 15. 3Sum


// Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that 
// i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

// Notice that the solution set must not contain duplicate triplets.


/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    
    let set = new Set();
    let low = 0;
    let high = nums.length -1;
    let mid = 1;
    
    nums.sort((a,b) => a-b)
    
    while(low < nums.length -3){
        if(nums[low] + nums[mid] + nums[high] === 0){
            set.add([nums[low], nums[mid], nums[high]])
        }else if(nums[low] + nums[mid] + nums[high] > 0){
            high--;
        }else{
            mid++;
        }
        
        if(mid >= high){
            low++
            high = nums.length - 1
            mid = low+1
        }
    }
    
    set.add([1,2])
    return Array.from(set);
};

File: python/15.py
# 15. 3Sum


# Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

# Notice that the solution set must not contain duplicate triplets.


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:

        ans = []
        nums.sort()
        a, l, r = 0, 1, len(nums) - 1

        while a < len(nums) - 2:
            while l < r:
                if nums[a] + nums[l] + nums[r] == 0:
                    # if [nums[a], nums[l], nums[r]] not in ans:
                    ans.append( [nums[a], nums[l], nums[r]] )
                    l += 1
                    while(nums[l] == nums[l-1] and l < r):
                        l+=1
                elif nums[a] + nums[l] + nums[r] < 0:
                    l += 1
                else:
                    r -= 1
            a += 1

            while 0 < a < len(nums)-2 and nums[a-1] == nums[a]:
                a += 1
            l = a + 1
            r = len(nums) -1
        return ans